### ISSUE TITLE
Allow options and flags after parameters

### ISSUE THREAD

**Issue by kke** [mdub/clamp#71]
Both should work:
`command delete --force filename.txt`
`command delete filename.txt --force`

Also this should work:

`command delete -- --filename.txt` (filename is "--filename.txt")

..and I think it already does:

```ruby
# option/parsing.rb:20
           break if switch == "--"
```

**Comment by uberjay** [comment]
This would be a very welcome addition! Thanks for working on this.

**Comment by mdub** [comment]
Oh boy. I distinctly recall writing a fairly long reply to this, on my phone, while on holiday in Europe, about a month ago.  But, there's no trace of it, which means you guys have been suffering radio silence from my end. Sorry about that.

Long story short, Clamp currently requires options to be specified after positional parameters (as you know), and that was an explicit design decision on my part. I don't think it's a good idea to allow options after parameters, because it would greatly complicate the parsing of the command-line.

I came up with some great examples where it (options after parameters) would cause problems, in the aforementioned long-reply-on-phone-from-Europe.  I'll need to try to recall what they were ...

**Comment by kke** [comment]
It's super annoying for users because they have to move cursor instead of just arrow-up + `--help` for example.

I already made the PR, it's just failing one test where it fails to display help correctly for `app PARAM subcommand` case.

**Comment by kke** [comment]
@SpComb clamp already recognizes and handles `--`

**Comment by uberjay** [comment]
Hrm. Yeah... this is (for me), one of the bigger usability issues with clamp. Because nearly every other tool *does* accept options after positional parameters, this makes for a somewhat jarring interactive experienced. (e.g., as @kke mentions, the quick up-arrrow & add an option to re-run a command doesn't work as expected.)

**Comment by mdub** [comment]
I _do_ get the "quick up-arrow and add an option" use-case. I understand that it's annoying to have to do that.

But, it's inaccurate to say that "nearly every other tool does accept options after positional parameters". My experience has been the opposite. Try these commands:

```
rm foo -f
ls ~/.ssh -l
git checkout new_branch -b
```

So, while _some_ command-line tools and CLI frameworks may allow options after arguments, it's not the norm, and I don't want it to be default behaviour in Clamp.

Several of my Clamp-based commands are _shims_ for other commands.  For example at `${CURRENT_WORKPLACE}` I've written a shim I'll call `as`, that executes commands with temporary AWS credentials, e.g.

```
as Admin aws ec2 describe-instances --region us-west-1
```

If Clamp looked for options anywhere on the command-line by default, `as` would not do the right thing. It would not be a backward-compatible change.

We can keep exploring the option (sic) of making options-after-arguments an opt-in feature, but I'm reluctant to make fundamental changes to the parsing logic.

**Comment by uberjay** [comment]
I'm not sure what you're trying to demonstrate with those three commands -- the first two work as expected for me, and the git example fails because `-b` takes a value.

**Comment by kke** [comment]
On mac `ls foo -al` doesn't work, one of my regular annoyances.

**Comment by uberjay** [comment]
Aw, that's terrible! I had no idea... one of the first things I do in macos is to add GNU coreutils to my PATH. d'oh...

Also, the git example that is similar to this would seems to works as expected:
```
$ git checkout HEAD^ -b new_branch
Switched to a new branch 'new_branch'
```

is equivalent to...
```
$ git checkout -b new_branch HEAD^
Switched to a new branch 'new_branch'
```

**Comment by uberjay** [comment]
For the record, I completely agree with you -- changing the default is a breaking change, and shouldn't be done lightly; however, having the *option* for this behavior would be pretty fantastic.

**Comment by mdub** [comment]
Fair point on my `git` example, @uberjay. But how about this one:

```
$ git diff HEAD^^ --help
usage: git diff [<options>] [<commit> [<commit>]] [--] [<path>...]
```

I'm surprised and somewhat nauseated that there are versions of `ls` and `rm` that allow options after arguments. Perhaps I'm just an old Unix fart, resistant to change.

**Comment by uberjay** [comment]
Yeah, it seems like this is a difference between GNU getopt and "standard" getopt. I suppose I'm not quite and old enough fart, having *mostly* used Linux with GNU userspace tools. This getopt behavior is a GNU extension. Apparently it's been the default behavior since glibc 1.09.1, which was released in 1996.

I think it's safe to say many people are going to be familiar with the non-standard GNU libc behavior of getopt. Aaaand, it's also pretty safe to say that many others (especially if they're avoided Linux!) will be familiar with the POSIX standard behavior.

And I also think changing the default behavior is not a great idea, but I'd really love it if it were an option. Or even if if the argument parsing were abstracted in a way that allowed me to provide an alternative parser.

**Comment by uberjay** [comment]
If you want to find more information about this, the magical keyword to search for is "permute":

> getopt is a system dependent function. The implementation of getopt in GNU C Library does permute the contents of the argument vector as it scans, so that eventually all the non-option arguments are at the end. On the contrary, the implementation of getopt in BSD C Library does not permute the argument vector and returns -1 if it encounters a non-option argument.

(from https://en.wikipedia.org/wiki/Getopt#History)

**Comment by mdub** [comment]
Thanks for the link, @uberjay. TIL.

I'm happy with making it possible to opt-in to the GNU `getopt(3)` behaviour.  Making it a class-level setting would make sense to me, e.g.

```ruby
Clamp do

  self.allow_options_after_arguments = true

  option "--loud", :flag, "say it loud"

  # ...

end
```

Possible alternatives to `allow_options_after_arguments = true`:
* `permute_arguments = true`
* `parse_like_a_gnu = true`
* `posixly_correct = false`

**Comment by uberjay** [comment]
That sounds good to me. I like your first suggestion, as well as `permute_arguments`. I think `posixly_correct` is too vague (i.e., there could be other ways in which any behavior is or is not posixly correct.)

I tried to find the equivalent option in my new favorite (non-ruby) argument parsing lib, `clap-rs`, but I *think* it doesn't have the ability to *disable* argument permuting. Or at least I couldn't find the flag for it. Looking in python's `argparse` module, they also don't appear to have a way to disable argument permuting. Granted, I haven't personally used the python library...

So, I think my vote is for `allow_options_after_arguments`. @kke -- perhaps you'd like to update your PR to conditionalize based on a flag? 😁

**Comment by kke** [comment]
I'd like if the PR passed the tests. I can update it if @mdub considers it worth it. There may be some better or more clampy way to do it than my attempt.

**Comment by mdub** [comment]
#76 released as `v1.2.0`.
